dhis-2/dhis-api/src/main/java/org/hisp/dhis/analytics/common/CommonParamsDelegator.java

```diff
@@ -0,0 +1,209 @@
+ * this list of conditions and the following disclaimer in the documentation
+ * and/or other materials provided with the distribution.
+ * Neither the name of the HISP project nor the names of its contributors may
+ * be used to endorse or promote products derived from this software without
+ * specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
+ * ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package org.hisp.dhis.analytics.common;
+
+import static java.util.Collections.emptyList;
+import static java.util.stream.Collectors.toList;
+import static java.util.stream.Collectors.toSet;
+import static org.hisp.dhis.common.DimensionalObject.ORGUNIT_DIM_ID;
+import static org.hisp.dhis.common.DimensionalObject.PERIOD_DIM_ID;
+
+import java.util.List;
+import java.util.Objects;
+import java.util.Set;
+
+import org.hisp.dhis.analytics.common.dimension.DimensionIdentifier;
+import org.hisp.dhis.analytics.common.dimension.DimensionParam;
+import org.hisp.dhis.common.BaseDimensionalObject;
+import org.hisp.dhis.common.DimensionItemKeywords.Keyword;
+import org.hisp.dhis.common.DimensionalItemObject;
+import org.hisp.dhis.common.DimensionalObject;
+import org.hisp.dhis.common.QueryItem;
+import org.hisp.dhis.legend.Legend;
+import org.hisp.dhis.option.Option;
+import org.hisp.dhis.program.ProgramStage;
+
+/**
+ * Component that provides useful methods based on the {@link CommonParams}
+ * object. Its main goal is to remove the responsibilities of
+ * {@link CommonParams} providing operations that highly depends on it (highly
+ * coupling).
+ *
+ * @author maikel arabori
+ */
+public class CommonParamsDelegator
+{
+    private final List<DimensionIdentifier<DimensionParam>> dimensionIdentifiers;
+
+    public CommonParamsDelegator( CommonParams commonParams )
+    {
+        this.dimensionIdentifiers = commonParams.getDimensionIdentifiers();
+    }
+
+    /**
+     * Retrieves all {@link Option} objects from all existing {@link QueryItem}
+     * objects part of the current "dimensionIdentifiers".
+     *
+     * @return the set of {@link Option}.
+     */
+    public Set<Option> getItemsOptions()
+    {
+        return dimensionIdentifiers.stream()
+            .filter( dimParam -> dimParam.getDimension() != null )
```

Comment:
in theory `dimensionParam.getDimension` will never be null, however I'd use `Objects::nonNull`